go/registry/api/sanity_check.go

```diff
@@ -147,43 +175,151 @@ func SanityCheckNodes(
+
+	// Generate escrow account for all entities.
+	for _, entity := range entities {
+		acct, ok := accounts[entity.ID]
+		if !ok {
+			return fmt.Errorf("registry: sanity check failed: no account associated with entity with ID: %v", entity.ID)
+		}
+
+		// Generate an escrow account with the same active balance and shares number.
+		escrow := &staking.EscrowAccount{
+			Active: staking.SharePool{
+				Balance:     acct.Escrow.Active.Balance,
+				TotalShares: acct.Escrow.Active.TotalShares,
+			},
+		}
+
+		// Add entity stake claim.
+		escrow.StakeAccumulator.AddClaimUnchecked(StakeClaimRegisterEntity, []staking.ThresholdKind{staking.KindEntity})
+
+		generatedEscrows[entity.ID] = escrow
+	}
+
+	for _, node := range nodes {
+		// Add node stake claims.
+		generatedEscrows[node.EntityID].StakeAccumulator.AddClaimUnchecked(StakeClaimForNode(node.ID), StakeThresholdsForNode(node))
+	}
+	for _, rt := range runtimes {
+		// Add runtime stake claims.
+		generatedEscrows[rt.EntityID].StakeAccumulator.AddClaimUnchecked(StakeClaimForRuntime(rt.ID), StakeThresholdsForRuntime(rt))
+	}
+
+	// Compare entities' generated escrow accounts with actual ones.
+	for _, entity := range entities {
+		generatedEscrow := generatedEscrows[entity.ID]
+		actualEscrow := accounts[entity.ID].Escrow
+
+		// Compare expected accumulator state with the actual one.
+		expectedClaims := generatedEscrows[entity.ID].StakeAccumulator.Claims
+		actualClaims := actualEscrow.StakeAccumulator.Claims
+		if len(expectedClaims) != len(actualClaims) {
+			return fmt.Errorf("incorrect number of stake claims for account %s (expected: %d got: %d)",
+				entity.ID,
+				len(expectedClaims),
+				len(actualClaims),
+			)
+		}
+		for claim, expectedThresholds := range expectedClaims {
+			thresholds, ok := actualClaims[claim]
+			if !ok {
+				return fmt.Errorf("missing claim %s for account %s", claim, entity.ID)
+			}
+			if len(thresholds) != len(expectedThresholds) {
+				return fmt.Errorf("incorrect number of thresholds for claim %s for account %s (expected: %d got: %d)",
+					claim,
+					entity.ID,
+					len(expectedThresholds),
+					len(thresholds),
+				)
+			}
+			for i, expectedThreshold := range expectedThresholds {
+				threshold := thresholds[i]
+				if threshold != expectedThreshold {
+					return fmt.Errorf("incorrect threshold in position %d for claim %s for account %s (expected: %s got: %s)",
+						i,
+						claim,
+						entity.ID,
+						expectedThreshold,
+						threshold,
+					)
+				}
+			}
+		}
+
+		// Check if entity has enough stake for all stake claims.
+		if err := generatedEscrow.CheckStakeClaims(stakeThresholds); err != nil {
+			var expected = "unknown"
+			expectedQty, err := generatedEscrow.StakeAccumulator.TotalClaims(stakeThresholds, nil)
+			if err == nil {
+				expected = expectedQty.String()
+			}
+			return fmt.Errorf("insufficient stake for account %s (expected: %s got: %s)", entity.ID, expected, generatedEscrow.Active.Balance)
```

Comment:
could we `%w` the error from `CheckStateClaims`?

or explain why we'd rather throw away that info

Reply:
Good catch. I've added `%w` to the `fmt.Errorf()` call.